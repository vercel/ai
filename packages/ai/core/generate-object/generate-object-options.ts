import { Schema } from '@ai-sdk/ui-utils';
import { z } from 'zod';
import { CallSettings } from '../prompt/call-settings';
import { Prompt } from '../prompt/prompt';
import { LanguageModel } from '../types/language-model';
import { ProviderMetadata } from '../types/provider-metadata';
import { ProviderOptions } from '../types/provider-metadata';
import { TelemetrySettings } from '../telemetry/telemetry-settings';
import { TypeValidationError } from '@ai-sdk/provider';
import { JSONParseError } from '@ai-sdk/provider';

/**
A function that attempts to repair the raw output of the mode
to enable JSON parsing.

Should return the repaired text or null if the text cannot be repaired.
     */
export type RepairTextFunction = (options: {
  text: string;
  error: JSONParseError | TypeValidationError;
}) => Promise<string | null>;

export type GenerateObjectBaseOptions = Omit<CallSettings, 'stopSequences'> &
  Prompt & {
    /**
The language model to use.
     */
    model: LanguageModel;

    /**
The mode to use for object generation.

The schema is converted into a JSON schema and used in one of the following ways

- 'auto': The provider will choose the best mode for the model.
- 'tool': A tool with the JSON schema as parameters is provided and the provider is instructed to use it.
- 'json': The JSON schema and an instruction are injected into the prompt. If the provider supports JSON mode, it iis enabled. If the provider supports JSON grammars, the grammar is used.

Please note that most providers do not support all modes.

Default and recommended: 'auto' (best mode for the model).
     */
    mode?: 'auto' | 'json' | 'tool';

    /**
A function that attempts to repair the raw output of the mode
to enable JSON parsing.
     */
    experimental_repairText?: RepairTextFunction;

    /**
Optional telemetry configuration (experimental).
 */

    experimental_telemetry?: TelemetrySettings;

    /**
Additional provider-specific options. They are passed through
to the provider from the AI SDK and enable provider-specific
functionality that can be fully encapsulated in the provider.
*/
    providerOptions?: ProviderOptions;

    /**
@deprecated Use `providerOptions` instead.
*/
    experimental_providerMetadata?: ProviderMetadata;

    /**
     * Internal. For test use only. May change without notice.
     */
    _internal?: {
      generateId?: () => string;
      currentDate?: () => Date;
    };
  };

/**
 * Schema-based options with common schema properties
 */
export interface SchemaBasedOptions<T> extends GenerateObjectBaseOptions {
  /**
   * The schema of the object that the model should generate.
   */
  schema: z.Schema<T, z.ZodTypeDef, any> | Schema<T>;

  /**
   * Optional name of the output that should be generated.
   * Used by some providers for additional LLM guidance, e.g.
   * via tool or schema name.
   */
  schemaName?: string;

  /**
   * Optional description of the output that should be generated.
   * Used by some providers for additional LLM guidance, e.g.
   * via tool or schema description.
   */
  schemaDescription?: string;
}

export type GenerateObjectNoSchemaOptions = GenerateObjectBaseOptions & {
  output: 'no-schema';
  /**
The mode to use for object generation. Must be "json" for no-schema output.
     */
  mode?: 'json';
};

export type GenerateObjectArrayOptions<ELEMENT> = GenerateObjectBaseOptions & {
  output: 'array';
  /**
The element schema of the array that the model should generate.
 */
  schema: z.Schema<ELEMENT, z.ZodTypeDef, any> | Schema<ELEMENT>;

  /**
Optional name of the array that should be generated.
Used by some providers for additional LLM guidance, e.g.
via tool or schema name.
*/
  schemaName?: string;

  /**
Optional description of the array that should be generated.
Used by some providers for additional LLM guidance, e.g.
via tool or schema description.
*/
  schemaDescription?: string;
};

export type GenerateObjectEnumOptions<ENUM extends string> =
  GenerateObjectBaseOptions & {
    output: 'enum';
    /**
The enum values that the model should use.
     */
    enum: Array<ENUM>;
  };

export type GenerateObjectObjectOptions<OBJECT> = GenerateObjectBaseOptions & {
  output?: 'object' | undefined;
  /**
The schema of the object that the model should generate.
     */
  schema: z.Schema<OBJECT, z.ZodTypeDef, any> | Schema<OBJECT>;

  /**
Optional name of the output that should be generated.
Used by some providers for additional LLM guidance, e.g.
via tool or schema name.
*/
  schemaName?: string;

  /**
Optional description of the output that should be generated.
Used by some providers for additional LLM guidance, e.g.
via tool or schema description.
*/
  schemaDescription?: string;
};

export type GenerateObjectOptions =
  | GenerateObjectNoSchemaOptions
  | GenerateObjectArrayOptions<any>
  | GenerateObjectEnumOptions<any>
  | GenerateObjectObjectOptions<any>;
