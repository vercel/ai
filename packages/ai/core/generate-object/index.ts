export { generateObject } from './generate-object';
export type { RepairTextFunction } from './generate-object-options';
export type { GenerateObjectResult } from './generate-object-result';
export { streamObject } from './stream-object';
export type { StreamObjectOnFinishCallback } from './stream-object';
export type {
  ObjectStreamPart,
  StreamObjectResult,
} from './stream-object-result';
export * from './generate-object-options';
