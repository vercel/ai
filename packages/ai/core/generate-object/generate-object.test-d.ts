import { expectTypeOf } from 'vitest';
import { generateObject } from './generate-object';
import { z } from 'zod';
import { JSONValue } from '@ai-sdk/provider';

describe('generateObject', () => {
  it('should support enum types', async () => {
    const result = await generateObject({
      output: 'enum',
      enum: ['a', 'b', 'c'] as const,
      model: undefined!,
    });

    expectTypeOf<typeof result.object>().toEqualTypeOf<'a' | 'b' | 'c'>;
  });

  it('should support schema types', async () => {
    const result = await generateObject({
      schema: z.object({ number: z.number() }),
      model: undefined!,
    });

    expectTypeOf<typeof result.object>().toEqualTypeOf<{ number: number }>();
  });

  it('should support no-schema output mode', async () => {
    const result = await generateObject({
      output: 'no-schema',
      model: undefined!,
    });

    expectTypeOf<typeof result.object>().toEqualTypeOf<JSONValue>();
  });

  it('should support array output mode', async () => {
    const result = await generateObject({
      output: 'array',
      schema: z.number(),
      model: undefined!,
    });

    expectTypeOf<typeof result.object>().toEqualTypeOf<number[]>();
  });
});
