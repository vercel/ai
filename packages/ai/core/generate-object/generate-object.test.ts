import { JSONParseError, TypeValidationError } from '@ai-sdk/provider';
import { jsonSchema } from '@ai-sdk/provider-utils';
import { convertReadableStreamToArray } from '@ai-sdk/provider-utils/test';
import assert, { fail } from 'node:assert';
import { z } from 'zod';
import { verifyNoObjectGeneratedError as originalVerifyNoObjectGeneratedError } from '../../src/error/no-object-generated-error';
import { MockLanguageModelV2 } from '../test/mock-language-model-v2';
import { MockTracer } from '../test/mock-tracer';
import { generateObject } from './generate-object';

const dummyResponseValues = {
  finishReason: 'stop' as const,
  usage: {
    inputTokens: 10,
    outputTokens: 20,
    totalTokens: 30,
    reasoningTokens: undefined,
    cachedInputTokens: undefined,
  },
  response: { id: 'id-1', timestamp: new Date(123), modelId: 'm-1' },
  warnings: [],
};

describe('output = "object"', () => {
  describe('result.object', () => {
    it('should generate object', async () => {
      const model = new MockLanguageModelV2({
        doGenerate: {
          ...dummyResponseValues,
          content: [{ type: 'text', text: '{ "content": "Hello, world!" }' }],
        },
      });
      const result = await generateObject({
        model,
        schema: z.object({ content: z.string() }),
        prompt: 'prompt',
      });

      expect(result.object).toMatchInlineSnapshot(`
        {
          "content": "Hello, world!",
        }
      `);
      expect(model.doGenerateCalls[0].prompt).toMatchInlineSnapshot(`
        [
          {
            "content": [
              {
                "text": "prompt",
                "type": "text",
              },
            ],
            "providerOptions": undefined,
            "role": "user",
          },
        ]
      `);
      expect(model.doGenerateCalls[0].responseFormat).toMatchInlineSnapshot(`
        {
          "description": undefined,
          "name": undefined,
          "schema": {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "additionalProperties": false,
            "properties": {
              "content": {
                "type": "string",
              },
            },
            "required": [
              "content",
            ],
            "type": "object",
          },
          "type": "json",
        }
      `);
    });

    it('should use name and description', async () => {
      const result = await generateObject({
        model: new MockLanguageModelV2({
          doGenerate: async ({ prompt, responseFormat }) => {
            expect(responseFormat).toStrictEqual({
              type: 'json',
              name: 'test-name',
              description: 'test description',
              schema: {
                $schema: 'http://json-schema.org/draft-07/schema#',
                additionalProperties: false,
                properties: { content: { type: 'string' } },
                required: ['content'],
                type: 'object',
              },
            });

            expect(prompt).toStrictEqual([
              {
                role: 'user',
                content: [{ type: 'text', text: 'prompt' }],
                providerOptions: undefined,
              },
            ]);

            return {
              ...dummyResponseValues,
              content: [
                { type: 'text', text: '{ "content": "Hello, world!" }' },
              ],
            };
          },
        }),
        schema: z.object({ content: z.string() }),
        schemaName: 'test-name',
        schemaDescription: 'test description',
        prompt: 'prompt',
      });

      assert.deepStrictEqual(result.object, { content: 'Hello, world!' });
    });
  });

  describe('result.request', () => {
    it('should contain request information', async () => {
      const result = await generateObject({
        model: new MockLanguageModelV2({
          doGenerate: async () => ({
            ...dummyResponseValues,
            content: [{ type: 'text', text: '{ "content": "Hello, world!" }' }],
            request: {
              body: 'test body',
            },
          }),
        }),
        schema: z.object({ content: z.string() }),
        prompt: 'prompt',
      });

      expect(result.request).toStrictEqual({
        body: 'test body',
      });
    });
  });

  describe('result.response', () => {
    it('should contain response information', async () => {
      const result = await generateObject({
        model: new MockLanguageModelV2({
          doGenerate: async () => ({
            ...dummyResponseValues,
            content: [{ type: 'text', text: '{ "content": "Hello, world!" }' }],
            response: {
              id: 'test-id-from-model',
              timestamp: new Date(10000),
              modelId: 'test-response-model-id',
              headers: {
                'custom-response-header': 'response-header-value',
              },
              body: 'test body',
            },
          }),
        }),
        schema: z.object({ content: z.string() }),
        prompt: 'prompt',
      });

      expect(result.response).toStrictEqual({
        id: 'test-id-from-model',
        timestamp: new Date(10000),
        modelId: 'test-response-model-id',
        headers: {
          'custom-response-header': 'response-header-value',
        },
        body: 'test body',
      });
    });
  });

  describe('zod schema', () => {
    it('should generate object when using zod transform', async () => {
      const model = new MockLanguageModelV2({
        doGenerate: {
          ...dummyResponseValues,
          content: [{ type: 'text', text: '{ "content": "Hello, world!" }' }],
        },
      });

      const result = await generateObject({
        model,
        schema: z.object({
          content: z.string().transform(value => value.length),
        }),
        prompt: 'prompt',
      });

      expect(result.object).toMatchInlineSnapshot(`
        {
          "content": 13,
        }
      `);
      expect(model.doGenerateCalls[0].prompt).toMatchInlineSnapshot(`
        [
          {
            "content": [
              {
                "text": "prompt",
                "type": "text",
              },
            ],
            "providerOptions": undefined,
            "role": "user",
          },
        ]
      `);
      expect(model.doGenerateCalls[0].responseFormat).toMatchInlineSnapshot(`
        {
          "description": undefined,
          "name": undefined,
          "schema": {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "additionalProperties": false,
            "properties": {
              "content": {
                "type": "string",
              },
            },
            "required": [
              "content",
            ],
            "type": "object",
          },
          "type": "json",
        }
      `);
    });

    it('should generate object when using zod prePreprocess', async () => {
      const model = new MockLanguageModelV2({
        doGenerate: {
          ...dummyResponseValues,
          content: [{ type: 'text', text: '{ "content": "Hello, world!" }' }],
        },
      });

      const result = await generateObject({
        model,
        schema: z.object({
          content: z.preprocess(
            val => (typeof val === 'number' ? String(val) : val),
            z.string(),
          ),
        }),
        prompt: 'prompt',
      });

      expect(result.object).toMatchInlineSnapshot(`
        {
          "content": "Hello, world!",
        }
      `);
      expect(model.doGenerateCalls[0].prompt).toMatchInlineSnapshot(`
        [
          {
            "content": [
              {
                "text": "prompt",
                "type": "text",
              },
            ],
            "providerOptions": undefined,
            "role": "user",
          },
        ]
      `);
      expect(model.doGenerateCalls[0].responseFormat).toMatchInlineSnapshot(`
        {
          "description": undefined,
          "name": undefined,
          "schema": {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "additionalProperties": false,
            "properties": {
              "content": {
                "type": "string",
              },
            },
            "required": [
              "content",
            ],
            "type": "object",
          },
          "type": "json",
        }
      `);
    });
  });

  describe('custom schema', () => {
    it('should generate object', async () => {
      const model = new MockLanguageModelV2({
        doGenerate: {
          ...dummyResponseValues,
          content: [{ type: 'text', text: '{ "content": "Hello, world!" }' }],
        },
      });

      const result = await generateObject({
        model,
        schema: jsonSchema({
          type: 'object',
          properties: { content: { type: 'string' } },
          required: ['content'],
          additionalProperties: false,
        }),
        prompt: 'prompt',
      });

      expect(result.object).toMatchInlineSnapshot(`
        {
          "content": "Hello, world!",
        }
      `);
      expect(model.doGenerateCalls[0].prompt).toMatchInlineSnapshot(`
        [
          {
            "content": [
              {
                "text": "prompt",
                "type": "text",
              },
            ],
            "providerOptions": undefined,
            "role": "user",
          },
        ]
      `);
      expect(model.doGenerateCalls[0].responseFormat).toMatchInlineSnapshot(`
        {
          "description": undefined,
          "name": undefined,
          "schema": {
            "additionalProperties": false,
            "properties": {
              "content": {
                "type": "string",
              },
            },
            "required": [
              "content",
            ],
            "type": "object",
          },
          "type": "json",
        }
      `);
    });
  });

  describe('result.toJsonResponse', () => {
    it('should return JSON response', async () => {
      const result = await generateObject({
        model: new MockLanguageModelV2({
          doGenerate: async ({}) => ({
            ...dummyResponseValues,
            content: [{ type: 'text', text: '{ "content": "Hello, world!" }' }],
          }),
        }),
        schema: z.object({ content: z.string() }),
        prompt: 'prompt',
      });

      const response = result.toJsonResponse();

      assert.strictEqual(response.status, 200);
      assert.strictEqual(
        response.headers.get('Content-Type'),
        'application/json; charset=utf-8',
      );

      assert.deepStrictEqual(
        await convertReadableStreamToArray(
          response.body!.pipeThrough(new TextDecoderStream()),
        ),
        ['{"content":"Hello, world!"}'],
      );
    });
  });

  describe('result.providerMetadata', () => {
    it('should contain provider metadata', async () => {
      const result = await generateObject({
        model: new MockLanguageModelV2({
          doGenerate: async ({}) => ({
            ...dummyResponseValues,
            content: [{ type: 'text', text: '{ "content": "Hello, world!" }' }],
            providerMetadata: {
              exampleProvider: {
                a: 10,
                b: 20,
              },
            },
          }),
        }),
        schema: z.object({ content: z.string() }),
        prompt: 'prompt',
      });

      expect(result.providerMetadata).toStrictEqual({
        exampleProvider: {
          a: 10,
          b: 20,
        },
      });
    });
  });

  describe('options.headers', () => {
    it('should pass headers to model', async () => {
      const result = await generateObject({
        model: new MockLanguageModelV2({
          doGenerate: async ({ headers }) => {
            expect(headers).toStrictEqual({
              'custom-request-header': 'request-header-value',
            });

            return {
              ...dummyResponseValues,
              content: [
                { type: 'text', text: '{ "content": "headers test" }' },
              ],
            };
          },
        }),
        schema: z.object({ content: z.string() }),
        prompt: 'prompt',
        headers: { 'custom-request-header': 'request-header-value' },
      });

      expect(result.object).toStrictEqual({ content: 'headers test' });
    });
  });

  describe('options.repairText', () => {
    it('should be able to repair a JSONParseError', async () => {
      const result = await generateObject({
        model: new MockLanguageModelV2({
          doGenerate: async ({}) => {
            return {
              ...dummyResponseValues,
              content: [
                {
                  type: 'text',
                  text: '{ "content": "provider metadata test" ',
                },
              ],
            };
          },
        }),
        schema: z.object({ content: z.string() }),
        prompt: 'prompt',
        experimental_repairText: async ({ text, error }) => {
          expect(error).toBeInstanceOf(JSONParseError);
          expect(text).toStrictEqual('{ "content": "provider metadata test" ');
          return text + '}';
        },
      });

      expect(result.object).toStrictEqual({
        content: 'provider metadata test',
      });
    });

    it('should be able to repair a TypeValidationError', async () => {
      const result = await generateObject({
        model: new MockLanguageModelV2({
          doGenerate: async ({}) => {
            return {
              ...dummyResponseValues,
              content: [
                {
                  type: 'text',
                  text: '{ "content-a": "provider metadata test" }',
                },
              ],
            };
          },
        }),
        schema: z.object({ content: z.string() }),
        prompt: 'prompt',
        experimental_repairText: async ({ text, error }) => {
          expect(error).toBeInstanceOf(TypeValidationError);
          expect(text).toStrictEqual(
            '{ "content-a": "provider metadata test" }',
          );
          return `{ "content": "provider metadata test" }`;
        },
      });

      expect(result.object).toStrictEqual({
        content: 'provider metadata test',
      });
    });

    it('should be able to handle repair that returns null', async () => {
      const result = generateObject({
        model: new MockLanguageModelV2({
          doGenerate: async ({}) => {
            return {
              ...dummyResponseValues,
              content: [
                {
                  type: 'text',
                  text: '{ "content-a": "provider metadata test" }',
                },
              ],
            };
          },
        }),
        schema: z.object({ content: z.string() }),
        prompt: 'prompt',
        experimental_repairText: async ({ text, error }) => {
          expect(error).toBeInstanceOf(TypeValidationError);
          expect(text).toStrictEqual(
            '{ "content-a": "provider metadata test" }',
          );
          return null;
        },
      });

      expect(result).rejects.toThrow(
        'No object generated: response did not match schema.',
      );
    });
  });

  describe('options.providerOptions', () => {
    it('should pass provider options to model', async () => {
      const result = await generateObject({
        model: new MockLanguageModelV2({
          doGenerate: async ({ providerOptions }) => {
            expect(providerOptions).toStrictEqual({
              aProvider: { someKey: 'someValue' },
            });

            return {
              ...dummyResponseValues,
              content: [
                {
                  type: 'text',
                  text: '{ "content": "provider metadata test" }',
                },
              ],
            };
          },
        }),
        schema: z.object({ content: z.string() }),
        prompt: 'prompt',
        providerOptions: {
          aProvider: { someKey: 'someValue' },
        },
      });

      expect(result.object).toStrictEqual({
        content: 'provider metadata test',
      });
    });
  });

  describe('error handling', () => {
    function verifyNoObjectGeneratedError(
      error: unknown,
      { message }: { message: string },
    ) {
      originalVerifyNoObjectGeneratedError(error, {
        message,
        response: {
          id: 'id-1',
          timestamp: new Date(123),
          modelId: 'm-1',
        },
        usage: {
          inputTokens: 10,
          outputTokens: 20,
          totalTokens: 30,
          reasoningTokens: undefined,
          cachedInputTokens: undefined,
        },
        finishReason: 'stop',
      });
    }

    it('should throw NoObjectGeneratedError when schema validation fails', async () => {
      try {
        await generateObject({
          model: new MockLanguageModelV2({
            doGenerate: async ({}) => ({
              ...dummyResponseValues,
              content: [{ type: 'text', text: '{ "content": 123 }' }],
            }),
          }),
          schema: z.object({ content: z.string() }),
          prompt: 'prompt',
        });

        fail('must throw error');
      } catch (error) {
        verifyNoObjectGeneratedError(error, {
          message: 'No object generated: response did not match schema.',
        });
      }
    });

    it('should throw NoObjectGeneratedError when parsing fails', async () => {
      try {
        await generateObject({
          model: new MockLanguageModelV2({
            doGenerate: async ({}) => ({
              ...dummyResponseValues,
              content: [{ type: 'text', text: '{ broken json' }],
            }),
          }),
          schema: z.object({ content: z.string() }),
          prompt: 'prompt',
        });

        fail('must throw error');
      } catch (error) {
        verifyNoObjectGeneratedError(error, {
          message: 'No object generated: could not parse the response.',
        });
      }
    });

    it('should throw NoObjectGeneratedError when parsing fails with repairResponse', async () => {
      try {
        await generateObject({
          model: new MockLanguageModelV2({
            doGenerate: async ({}) => ({
              ...dummyResponseValues,
              content: [{ type: 'text', text: '{ broken json' }],
            }),
          }),
          schema: z.object({ content: z.string() }),
          prompt: 'prompt',
          experimental_repairText: async ({ text }) => text + '{',
        });

        fail('must throw error');
      } catch (error) {
        verifyNoObjectGeneratedError(error, {
          message: 'No object generated: could not parse the response.',
        });
      }
    });

    it('should throw NoObjectGeneratedError when no text is available', async () => {
      try {
        await generateObject({
          model: new MockLanguageModelV2({
            doGenerate: async ({}) => ({
              ...dummyResponseValues,
              content: [],
            }),
          }),
          schema: z.object({ content: z.string() }),
          prompt: 'prompt',
        });

        fail('must throw error');
      } catch (error) {
        verifyNoObjectGeneratedError(error, {
          message: 'No object generated: the model did not return a response.',
        });
      }
    });
  });
});

describe('output = "array"', () => {
  it('should generate an array with 3 elements', async () => {
    const model = new MockLanguageModelV2({
      doGenerate: {
        ...dummyResponseValues,
        content: [
          {
            type: 'text',
            text: JSON.stringify({
              elements: [
                { content: 'element 1' },
                { content: 'element 2' },
                { content: 'element 3' },
              ],
            }),
          },
        ],
      },
    });

    const result = await generateObject({
      model,
      schema: z.object({ content: z.string() }),
      output: 'array',
      prompt: 'prompt',
    });

    expect(result.object).toMatchInlineSnapshot(`
      [
        {
          "content": "element 1",
        },
        {
          "content": "element 2",
        },
        {
          "content": "element 3",
        },
      ]
    `);
    expect(model.doGenerateCalls[0].prompt).toMatchInlineSnapshot(`
      [
        {
          "content": [
            {
              "text": "prompt",
              "type": "text",
            },
          ],
          "providerOptions": undefined,
          "role": "user",
        },
      ]
    `);
    expect(model.doGenerateCalls[0].responseFormat).toMatchInlineSnapshot(`
      {
        "description": undefined,
        "name": undefined,
        "schema": {
          "$schema": "http://json-schema.org/draft-07/schema#",
          "additionalProperties": false,
          "properties": {
            "elements": {
              "items": {
                "additionalProperties": false,
                "properties": {
                  "content": {
                    "type": "string",
                  },
                },
                "required": [
                  "content",
                ],
                "type": "object",
              },
              "type": "array",
            },
          },
          "required": [
            "elements",
          ],
          "type": "object",
        },
        "type": "json",
      }
    `);
  });
});

describe('output = "enum"', () => {
  it('should generate an enum value', async () => {
    const model = new MockLanguageModelV2({
      doGenerate: {
        ...dummyResponseValues,
        content: [
          {
            type: 'text',
            text: JSON.stringify({ result: 'sunny' }),
          },
        ],
      },
    });

    const result = await generateObject({
      model,
      output: 'enum',
      enum: ['sunny', 'rainy', 'snowy'],
      prompt: 'prompt',
    });

    expect(result.object).toEqual('sunny');
    expect(model.doGenerateCalls[0].prompt).toMatchInlineSnapshot(`
      [
        {
          "content": [
            {
              "text": "prompt",
              "type": "text",
            },
          ],
          "providerOptions": undefined,
          "role": "user",
        },
      ]
    `);
    expect(model.doGenerateCalls[0].responseFormat).toMatchInlineSnapshot(`
      {
        "description": undefined,
        "name": undefined,
        "schema": {
          "$schema": "http://json-schema.org/draft-07/schema#",
          "additionalProperties": false,
          "properties": {
            "result": {
              "enum": [
                "sunny",
                "rainy",
                "snowy",
              ],
              "type": "string",
            },
          },
          "required": [
            "result",
          ],
          "type": "object",
        },
        "type": "json",
      }
    `);
  });
});

describe('output = "no-schema"', () => {
  it('should generate object', async () => {
    const model = new MockLanguageModelV2({
      doGenerate: {
        ...dummyResponseValues,
        content: [{ type: 'text', text: '{ "content": "Hello, world!" }' }],
      },
    });

    const result = await generateObject({
      model,
      output: 'no-schema',
      prompt: 'prompt',
    });

    expect(result.object).toMatchInlineSnapshot(`
      {
        "content": "Hello, world!",
      }
    `);
    expect(model.doGenerateCalls[0].prompt).toMatchInlineSnapshot(`
      [
        {
          "content": [
            {
              "text": "prompt",
              "type": "text",
            },
          ],
          "providerOptions": undefined,
          "role": "user",
        },
      ]
    `);
    expect(model.doGenerateCalls[0].responseFormat).toMatchInlineSnapshot(`
      {
        "description": undefined,
        "name": undefined,
        "schema": undefined,
        "type": "json",
      }
    `);
  });
});

describe('telemetry', () => {
  let tracer: MockTracer;

  beforeEach(() => {
    tracer = new MockTracer();
  });

  it('should not record any telemetry data when not explicitly enabled', async () => {
    await generateObject({
      model: new MockLanguageModelV2({
        doGenerate: async () => ({
          ...dummyResponseValues,
          content: [{ type: 'text', text: '{ "content": "Hello, world!" }' }],
        }),
      }),
      schema: z.object({ content: z.string() }),
      prompt: 'prompt',
    });

    assert.deepStrictEqual(tracer.jsonSpans, []);
  });

  it('should record telemetry data when enabled', async () => {
    await generateObject({
      model: new MockLanguageModelV2({
        doGenerate: async () => ({
          ...dummyResponseValues,
          content: [{ type: 'text', text: '{ "content": "Hello, world!" }' }],
          response: {
            id: 'test-id-from-model',
            timestamp: new Date(10000),
            modelId: 'test-response-model-id',
          },
        }),
      }),
      schema: z.object({ content: z.string() }),
      schemaName: 'test-name',
      schemaDescription: 'test description',
      prompt: 'prompt',
      topK: 0.1,
      topP: 0.2,
      frequencyPenalty: 0.3,
      presencePenalty: 0.4,
      temperature: 0.5,
      headers: {
        header1: 'value1',
        header2: 'value2',
      },
      experimental_telemetry: {
        isEnabled: true,
        functionId: 'test-function-id',
        metadata: {
          test1: 'value1',
          test2: false,
        },
        tracer,
      },
    });

    expect(tracer.jsonSpans).toMatchSnapshot();
  });

  it('should not record telemetry inputs / outputs when disabled', async () => {
    await generateObject({
      model: new MockLanguageModelV2({
        doGenerate: async () => ({
          ...dummyResponseValues,
          content: [{ type: 'text', text: '{ "content": "Hello, world!" }' }],
          response: {
            id: 'test-id-from-model',
            timestamp: new Date(10000),
            modelId: 'test-response-model-id',
          },
        }),
      }),
      schema: z.object({ content: z.string() }),
      prompt: 'prompt',
      experimental_telemetry: {
        isEnabled: true,
        recordInputs: false,
        recordOutputs: false,
        tracer,
      },
    });

    expect(tracer.jsonSpans).toMatchSnapshot();
  });
});

describe('options.messages', () => {
  it('should support models that use "this" context in supportedUrls', async () => {
    let supportedUrlsCalled = false;
    class MockLanguageModelWithImageSupport extends MockLanguageModelV2 {
      constructor() {
        super({
          supportedUrls: () => {
            supportedUrlsCalled = true;
            // Reference 'this' to verify context
            return this.modelId === 'mock-model-id'
              ? ({ 'image/*': [/^https:\/\/.*$/] } as Record<string, RegExp[]>)
              : {};
          },
          doGenerate: async () => ({
            ...dummyResponseValues,
            content: [{ type: 'text', text: '{ "content": "Hello, world!" }' }],
          }),
        });
      }
    }

    const model = new MockLanguageModelWithImageSupport();

    const result = await generateObject({
      model,
      schema: z.object({ content: z.string() }),
      messages: [
        {
          role: 'user',
          content: [{ type: 'image', image: 'https://example.com/test.jpg' }],
        },
      ],
    });

    expect(result.object).toStrictEqual({ content: 'Hello, world!' });
    expect(supportedUrlsCalled).toBe(true);
  });
});
