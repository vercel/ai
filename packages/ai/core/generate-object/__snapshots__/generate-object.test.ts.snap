// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`telemetry > should not record telemetry inputs / outputs when disabled 1`] = `
[
  {
    "attributes": {
      "ai.model.id": "mock-model-id",
      "ai.model.provider": "mock-provider",
      "ai.operationId": "ai.generateObject",
      "ai.response.finishReason": "stop",
      "ai.settings.maxRetries": 2,
      "ai.settings.output": "object",
      "ai.usage.completionTokens": 20,
      "ai.usage.promptTokens": 10,
      "operation.name": "ai.generateObject",
    },
    "events": [],
    "name": "ai.generateObject",
  },
  {
    "attributes": {
      "ai.model.id": "mock-model-id",
      "ai.model.provider": "mock-provider",
      "ai.operationId": "ai.generateObject.doGenerate",
      "ai.response.finishReason": "stop",
      "ai.response.id": "test-id-from-model",
      "ai.response.model": "test-response-model-id",
      "ai.response.timestamp": "1970-01-01T00:00:10.000Z",
      "ai.settings.maxRetries": 2,
      "ai.usage.completionTokens": 20,
      "ai.usage.promptTokens": 10,
      "gen_ai.request.model": "mock-model-id",
      "gen_ai.response.finish_reasons": [
        "stop",
      ],
      "gen_ai.response.id": "test-id-from-model",
      "gen_ai.response.model": "test-response-model-id",
      "gen_ai.system": "mock-provider",
      "gen_ai.usage.input_tokens": 10,
      "gen_ai.usage.output_tokens": 20,
      "operation.name": "ai.generateObject.doGenerate",
    },
    "events": [],
    "name": "ai.generateObject.doGenerate",
  },
]
`;

exports[`telemetry > should record telemetry data when enabled 1`] = `
[
  {
    "attributes": {
      "ai.model.id": "mock-model-id",
      "ai.model.provider": "mock-provider",
      "ai.operationId": "ai.generateObject",
      "ai.prompt": "{"prompt":"prompt"}",
      "ai.request.headers.header1": "value1",
      "ai.request.headers.header2": "value2",
      "ai.response.finishReason": "stop",
      "ai.response.object": "{"content":"Hello, world!"}",
      "ai.schema": "{"type":"object","properties":{"content":{"type":"string"}},"required":["content"],"additionalProperties":false,"$schema":"http://json-schema.org/draft-07/schema#"}",
      "ai.schema.description": "test description",
      "ai.schema.name": "test-name",
      "ai.settings.frequencyPenalty": 0.3,
      "ai.settings.maxRetries": 2,
      "ai.settings.output": "object",
      "ai.settings.presencePenalty": 0.4,
      "ai.settings.temperature": 0.5,
      "ai.settings.topK": 0.1,
      "ai.settings.topP": 0.2,
      "ai.telemetry.functionId": "test-function-id",
      "ai.telemetry.metadata.test1": "value1",
      "ai.telemetry.metadata.test2": false,
      "ai.usage.completionTokens": 20,
      "ai.usage.promptTokens": 10,
      "operation.name": "ai.generateObject test-function-id",
      "resource.name": "test-function-id",
    },
    "events": [],
    "name": "ai.generateObject",
  },
  {
    "attributes": {
      "ai.model.id": "mock-model-id",
      "ai.model.provider": "mock-provider",
      "ai.operationId": "ai.generateObject.doGenerate",
      "ai.prompt.messages": "[{"role":"user","content":[{"type":"text","text":"prompt"}]}]",
      "ai.request.headers.header1": "value1",
      "ai.request.headers.header2": "value2",
      "ai.response.finishReason": "stop",
      "ai.response.id": "test-id-from-model",
      "ai.response.model": "test-response-model-id",
      "ai.response.object": "{ "content": "Hello, world!" }",
      "ai.response.timestamp": "1970-01-01T00:00:10.000Z",
      "ai.settings.frequencyPenalty": 0.3,
      "ai.settings.maxRetries": 2,
      "ai.settings.presencePenalty": 0.4,
      "ai.settings.temperature": 0.5,
      "ai.settings.topK": 0.1,
      "ai.settings.topP": 0.2,
      "ai.telemetry.functionId": "test-function-id",
      "ai.telemetry.metadata.test1": "value1",
      "ai.telemetry.metadata.test2": false,
      "ai.usage.completionTokens": 20,
      "ai.usage.promptTokens": 10,
      "gen_ai.request.frequency_penalty": 0.3,
      "gen_ai.request.model": "mock-model-id",
      "gen_ai.request.presence_penalty": 0.4,
      "gen_ai.request.temperature": 0.5,
      "gen_ai.request.top_k": 0.1,
      "gen_ai.request.top_p": 0.2,
      "gen_ai.response.finish_reasons": [
        "stop",
      ],
      "gen_ai.response.id": "test-id-from-model",
      "gen_ai.response.model": "test-response-model-id",
      "gen_ai.system": "mock-provider",
      "gen_ai.usage.input_tokens": 10,
      "gen_ai.usage.output_tokens": 20,
      "operation.name": "ai.generateObject.doGenerate test-function-id",
      "resource.name": "test-function-id",
    },
    "events": [],
    "name": "ai.generateObject.doGenerate",
  },
]
`;
